DEP: Deprecate (rather than remove) the int-via-float parsing in loadtxt

seberg opened the pull request:
Note that previously `float()` was used, so that we still lose the
capability of parsing underscores.  I think we should just live with
that.
This is mainly to allow a smoother transition, since:

   from skimage import feature

happens to run into it (and other libraries might as well in principle)

---

IMO, this should be good enough to close the issue.  I do think we should go with this as a deprecation right now.  **This is draft, since I would like a link to a brief docs page.** @rossbar do you think you may have time to look into that?  Otherwise, I will think about it a bit.

~(Whoops, plus tests are still a TODO at this time :), but will do that right away.).~

Closes gh-21652

seberg committed:
DEP: Deprecate (rather than remove) the int-via-float parsing in loadtxt

Note that previously `float()` was used, so that we still lose the
capability of parsing underscores.  I think we should just live with
that.
This is mainly to allow a smoother transition, since:

   from skimage import feature

happens to run into it (and other libraries might as well in principle)

seberg committed:
TST: Add deprecation testcase for int-via-float in loadtxt

seberg commented on numpy/core/src/multiarray/textreading/str_to_int.c:
```diff
@@ -41,8 +63,24 @@
+                }                                                           \
+            }                                                               \
+            pconfig->gave_int_via_float_warning = true;                     \
+            x = (uintw##_t)fval;                                            \
```
Now that we have C++, this should even more be cleaned up and stop being a macro.  But I don't think it is important enough to try that now.  Especially on something that needs backporting.

rossbar commented:
>  @rossbar do you think you may have time to look into that?

Yes, though I'll be out for the next few days! Don't wait on me for anything time-sensitive, I can definitely look when I'm back at the computer

seberg committed:
DOC: Update the loadtxt int-via-float deprecation warning message

Include brief instructions directly into the warning message rather
than trying to write a long document as part of this PR.
A long document may still be nice (but maybe not necessary)

charris commented:
The failing test can be ignored.

> Should we merge this, and I create a backport PR directly editing the release note?

Sounds good. I'll put this in now. Thanks Sebastian.

hmaarrfk commented:
Thanks all!
